### PR TITLE
[REQUEST FOR REVIEW] [MXNET-109] Logging APIs for Visualizing MXNet Data in TensorBoard

### PR DESCRIPTION
**Pull request by reminisce**
## Description ##
This PR implements Python APIs of logging MXNet data for visualization in [TensorBoard](https://www.tensorflow.org/programmers_guide/summaries_and_tensorboard).
- Design proposal: https://cwiki.apache.org/confluence/display/MXNET/Logging+MXNet+Data+for+Visualization+in+TensorBoard
- Jira link: https://issues.apache.org/jira/browse/MXNET-109

### Why choose TensorBoard as the frontend for data visualization? ###
- TensorBoard has a nice GUI for data visualization and it's production environment friendly.
- Installing TensorBoard is as simple as typing `pip install tensorboard` in a terminal.

### How are MXNet logging APIs implemented? ###
- The major logging components are ported from their Python and C++ implementations in [TensorFlow](https://github.com/tensorflow/tensorflow) as documented in the code.
- The algorithm of encoding data written to event files are borrowed from https://github.com/TeamHG-Memex/tensorboard_logger.
- We adopted the user-level logging API design from https://github.com/lanpa/tensorboard-pytorch.
- This work is inspired by @zihaolucky 's pioneering efforts on [dmlc/tensorboard](https://github.com/dmlc/tensorboard). Special thanks to him for many technical discussions during the process of development.

### How to use MXNet logging APIs? ###
Please find the details in the following document: https://github.com/reminisce/mxnet/blob/tensorboard_logging/docs/api/python/contrib/summary.md

### Is there an example demonstrating logging data for visualization in a training job? ###
Yes, please see here for logging data when training an MNIST model using the Gluon interfaces.
https://github.com/reminisce/mxnet/blob/tensorboard_logging/example/visualization/tensorboard/mnist.py. Search for `sw` (short for `SummaryWriter`) in the code to see how logging APIs are used.

## References ##
1. https://github.com/dmlc/tensorboard
2. https://github.com/TeamHG-Memex/tensorboard_logger
3. https://github.com/lanpa/tensorboard-pytorch
4. https://github.com/tensorflow/tensorflow

Please help to review and feel free to @ more reviewers. Thanks!
@piiswrong @szha @eric-haibin-lin @zhreshold @astonzhang @zihaolucky @cjolivier01 @marcoabreu @anirudh2290 @rahul003 @haojin2 

### Essentials ###
- [ ] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

File: docs/api/python/contrib/summary.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
+# Logging MXNet Data for Visualization in TensorBoard
```
Comment (marcoabreu) [inline docs/api/python/contrib/summary.md]: Is this file location the right place? I would have expected it to be at python/contrib/summary/README.md

File: example/visualization/tensorboard/mnist.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        print('[Epoch %d] Validation: %s=%f' % (epoch, name, val_acc))
+        sw.add_scalar(tag='valid_acc', value=val_acc, global_step=epoch)
+
+    net.save_params('mnist.params')
+    sw.close()
+
+
+if __name__ == '__main__':
+    if opt.cuda:
+        ctx = mx.gpu(0)
```
Comment (marcoabreu) [inline example/visualization/tensorboard/mnist.py]: Would this limit the training to only 1 gpu?

File: tests/python/gpu/test_operator_gpu.py [hunk 3]
```diff
@@ -45,6 +45,7 @@
 from test_sparse_ndarray import test_sparse_nd_setitem, test_sparse_nd_binary_scalar_op
 from test_sparse_operator import *
 from test_ndarray import *
+from test_summary import *
```
Comment (marcoabreu) [inline tests/python/gpu/test_operator_gpu.py]: Does this really have to run on a GPU? In general, this seems like an integration test to me - I think it would be better to have it in the Integration stage of our Jenkins pipeline

File: tests/python/unittest/test_summary.py [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+import shutil
+from mxnet.contrib.summary import SummaryWriter
+from mxnet.contrib.summary.utils import _make_metadata_tsv, make_image_grid, _make_sprite_image
+from mxnet.contrib.summary.utils import _add_embedding_config, _save_embedding_tsv
+from mxnet.test_utils import *
+from common import with_seed
+
+# DO NOT CHANGE THESE NAMES AS THEY FOLLOW THE DEFINITIONS IN TENSORBOARD
+_LOGDIR = './logs_for_tensorboard'
```
Comment (marcoabreu) [inline tests/python/unittest/test_summary.py]: Could you use a tempdir? https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/common.py#L231

File: example/visualization/tensorboard/mnist.py (hunk 2 above)
Comment (reminisce) [reply example/visualization/tensorboard/mnist.py]: No, it would not.

File: tests/python/gpu/test_operator_gpu.py (hunk 3 above)
Comment (reminisce) [reply tests/python/gpu/test_operator_gpu.py]: Yes, because we need to verify the logging APIs can take care of `NDArray`s allocated on GPUs as correctly as on CPUs. Logging is just the final step for those APIs. There are also data processing algorithms implemented for different APIs.

File: tests/python/unittest/test_summary.py (hunk 4 above)
Comment (cjolivier01) [reply tests/python/unittest/test_summary.py]: one problem with using tempdir is that for troubleshooting, no one will know where to look for logs without looking in the code.

Comment (marcoabreu) [reply tests/python/unittest/test_summary.py]: Good point. How about we introduce a temp-dir in the root of mxnet that is covered by gitignore? At the moment, running tests will result in a lot of files being created.

Comment (marcoabreu) [reply tests/python/unittest/test_summary.py]: Yes, can we put it in a local temp anyways? E.g. ./TMP/model1.json

File: python/setup.py [hunk 5]
```diff
@@ -99,6 +99,18 @@ def config_cython():
...
 
 
+def compile_summary_protobuf():
+    proto_path = '../python/mxnet/contrib/summary/proto'
+    proto_files = os.path.join(proto_path, '*.proto')
+    cmd = 'protoc ' + proto_files + ' --python_out=' + proto_path + ' --proto_path=' + proto_path
+    return os.system(cmd)
+
+
+if compile_summary_protobuf() != 0:
```
Comment (reminisce) [inline python/setup.py]: @marcoabreu Does CI run this file for installing the MXNet python package? The current unit tests are failing because the protobuf files are not compiled into `.py` files as indicated in this file before installation. How could I configure the CI to compile the protobuf files first?

File: tests/python/unittest/test_summary.py (hunk 4 above)
Comment (cjolivier01) [reply tests/python/unittest/test_summary.py]: actually I’m not crazy about a “temp sir”, and I’ll tell you why — try not to look at this from a standpoint of mxnet being an app, because it isn’t.  It’s a library. It runs in whatever directory is the current directory for which it is invoked. you have no idea what that directory may be, and it’s quite possible that whoever is using the library doesn’t have any idea about the logging mechanism and expecting to create a directory TMP on a unix filesystem is a stretch in presumption.  there could be an environment var override, possibly, but it’s easiest for everyone to find it to just dump it in cwd. profiler for example goes to current dir by default.
*.log can be added to .gitignore

File: docs/api/python/contrib/summary.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+The module `mxnet.contrib.summary` enables MXNet users to visualize data in
+[TensorBoard](https://www.tensorflow.org/programmers_guide/summaries_and_tensorboard). 
+Please note that this module only provides the APIs for data logging. For visualization,
+users still need to install TensorBoard.
+
+### How to install TensorBoard
+To launch TensorBoard for visualization, make sure you have the
+[official release of TensorBoard](https://pypi.python.org/pypi/tensorboard) installed.
+You can type `pip install tensorboard` on you machine to install TensorBoard.
```
Comment (fhieber) [inline docs/api/python/contrib/summary.md]: Small comment: the official Tensorboard package requires Tensorflow, but does not declare this dependency in the pypi package (https://pypi.python.org/pypi/tensorboard). So you might need to add a 'pip install tensorflow' here.

File: docs/api/python/contrib/summary.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+### How to launch TensorBoard
+After you installed the TensorBoar Python package, type the following command in the terminal
+to launch TensorBoard:
+```
+tensorborad --logdir=/path/to/your/log/dir --host=your_host_ip --port=your_port_number
```
Comment (fhieber) [inline docs/api/python/contrib/summary.md]: typo here and in line 23: tensorborad -> tensorboard

File: docs/api/python/contrib/summary.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+y_vals = np.sin(x_vals)
+with SummaryWriter(logdir='./logs') as sw:
+    for x, y in zip(x_vals, y_vals):
+        sw.add_scalar(tag='sin_function_curve', value=y, global_step=x * 100)
+```
+![png](https://github.com/reminisce/web-data/blob/tensorboard_doc/mxnet/tensorboard/doc/summary_scalar_sin.png)
+
+
+### Histogram
+We can visulize the value distributions of tensors by logging `NDArray`s in terms of histograms.
```
Comment (fhieber) [inline docs/api/python/contrib/summary.md]: typo: visualize

File: python/setup.py (hunk 5 above)
Comment (marcoabreu) [reply python/setup.py]: No it doesn't.

File: docs/api/python/contrib/summary.md (hunk 6 above)
Comment (reminisce) [reply docs/api/python/contrib/summary.md]: You are right. Thanks for such a careful and thorough proofreading on the documentation. I will address all the comments as suggested. :)

Comment (piiswrong) [reply docs/api/python/contrib/summary.md]: Do we have to make user install tf? how about zihao's extracted tensorboard?

File: docs/api/python/contrib/summary.md (hunk 1 above)
Comment (piiswrong) [inline docs/api/python/contrib/summary.md]: Do we want to call it summary? Should the name be something that explicitly points to tensorboard?

File: python/mxnet/contrib/summary/crc32c.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (piiswrong) [inline python/mxnet/contrib/summary/crc32c.py]: _crc32c.py to hide it?

**Comment by mli** [comment]
How we compile these proto files? Does tensorboard include protobuf as the requirement? I'm not sure if it will conflict with ps-lite's statically linked protobuf.

**Comment by reminisce** [comment]
@mli 
1. The protobuf files are compiled when users type `python setup.py install` under `python/` directory. See here: https://github.com/reminisce/mxnet/blob/4a45b2cd01f0723813d83b9f5f5cdd2a70843184/python/setup.py#L109
2. TensorBoard requires protobuf. Before writing data to event files, they are first transferred to the corresponding protobuf definitions, then serialized to binary strings and encoded in the format of `tf.record`, and finally written to event files.

**Comment by mli** [comment]
I also have a complicated feeling of having a protobuf dependency. What current users need to do to use it?

1. install protobuf
2. install mxnet with proto compiled (hope we can enable it on pip). 
3. install tensorboard to visualize the results.

having a separate module may be more straightforward

1. install mxnet as normal
2. install tensorboard (i guess it will install protobuf together?)
3. install mxnet-tensorboard

by this, we assume people will visualize the results on the same machine. otherwise,

1. install tensorboard on local machine
2. install mxnet, and mxnet-tensorboard on work machines, assume mxnet-tensorboard will have protobuf as the dependency

**Comment by rahul003** [comment]
I was trying to use an example long back with tensorboard and I remember running into issues because of different protobuf versions required for tensorboard and mxnet. 
@reminisce Could we check if this is an issue by compiling with USE_DIST_KVSTORE and running some example? I imagine the version used by tensorboard is already newer (mxnet uses 2.5.0).

**Comment by reminisce** [comment]
@rahul003 I think the problem is that [meta.proto](https://github.com/dmlc/ps-lite/blob/aee325276bccb092f516df0bce30d3a8333f4038/src/meta.proto) in ps-lite is protobuf2 while tensorboard's protobuf is version 3. Did you compile those protobuf files using the their corresponding version of compilers?

File: tests/python/unittest/test_summary.py (hunk 4 above)
Comment (marcoabreu) [reply tests/python/unittest/test_summary.py]: Very good point, Chris! This sounds reasonable and seems to be a good solution.

File: tests/python/gpu/test_operator_gpu.py (hunk 3 above)
Comment (marcoabreu) [reply tests/python/gpu/test_operator_gpu.py]: Great, thanks

File: python/setup.py (hunk 5 above)
Comment (marcoabreu) [reply python/setup.py]: As explained offline, we're not running setup.py as part of our CI pipeline - instead, we're defining the path to the mxnet lib as an environment variable. I'd appreciate it you could integrate setup.py and remove that env var.

**Comment by cjolivier01** [comment]
what’s stopping us from bringing mxnet/ps-lite into the modern age and using the same protobuf version?

**Comment by reminisce** [comment]
As discussed offline, we are going to moving the implementation to a standalone repo.

File: docs/api/python/contrib/summary.md (hunk 6 above)
Comment (zihaolucky) [reply docs/api/python/contrib/summary.md]: If we use it in AMI, then TF is already installed and everything is fine.

**Comment by zihaolucky** [comment]
@reminisce We can move to a new repo and build the PyPI package as we do in dmlc/tensorboard, it would compile all protos and we just add this package as the dependency in MXNet.

**Comment by reminisce** [comment]
@zihaolucky That's what we are working on right now. Thanks for the suggestion. :)
